### PR TITLE
Sync application resource, when an application resource on argocd namespace is deleted.

### PR DESCRIPTION
**Pull request by zoetrope**
Reconcile process is not triggered when an application resource on argocd namespace is deleted.
Fix it to queue the tenant application when it detects the deletion of an application resource.

Signed-off-by: zoetrope <a.ikezoe@gmail.com>

### CODE REVIEW

File: controllers/application_controller.go [hunk 1]
```diff
@@ -50,16 +54,34 @@ type ApplicationReconciler struct {
...
-	if err := r.client.Get(ctx, req.NamespacedName, app); err != nil {
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+
+	err := r.client.Get(ctx, req.NamespacedName, app)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	}
+
+	if apierrors.IsNotFound(err) && req.Namespace == r.config.ArgoCD.Namespace {
+		logger.Info("argocd application was deleted")
```
Comment (yokaze) [inline controllers/application_controller.go:64]: It seems good to write the application name.

File: controllers/application_controller.go [hunk 2]
```diff
@@ -83,7 +115,12 @@ func (r *ApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		if apierrors.IsNotFound(err) {
 			argocdApp = nil
 		}
-
+		if argocdApp != nil && argocdApp.GetDeletionTimestamp() != nil {
+			logger.Info("argocd application is deleting. requeue the request")
+			return ctrl.Result{
+				Requeue: true,
```
Comment (yokaze) [inline controllers/application_controller.go:121]: This is polling, I think it's better to use RequeueAfter with some waiting duration.

File: controllers/application_controller.go [hunk 3]
```diff
@@ -132,6 +169,7 @@ func (r *ApplicationReconciler) reconcileApplication(ctx context.Context, argocd
 		return ctrl.Result{}, err
 	}
 	if !canSync {
+		logger.Info("cannot sync the application spec. requeue the request after 30 minutes")
```
Comment (yokaze) [inline controllers/application_controller.go:172]: Clarifying the application name is better.

File: controllers/application_controller.go [hunk 4]
```diff
@@ -149,6 +187,7 @@ func (r *ApplicationReconciler) reconcileApplication(ctx context.Context, argocd
 		logger.Error(err, "failed to sync application status")
 		return ctrl.Result{}, err
 	}
+	logger.Info("Application successfully reconciled")
```
Comment (yokaze) [inline controllers/application_controller.go:190]: Clarifying the application name is better; or it would be good to be a debug level message.

**Review by yokaze — CHANGES_REQUESTED** [review]

File: controllers/application_controller.go (hunk 1 above)
Comment (zoetrope) [reply controllers/application_controller.go:64]: I don't think it is necessary.
In reconcile function, the logger includes `name` and `namespace`.

https://github.com/kubernetes-sigs/controller-runtime/blob/e42a8c17945360813e64bde2c4ff742515b750ba/pkg/builder/controller.go#L325
https://zenn.dev/zoetro/articles/760346baab7e24#controller-runtime%E3%81%8C%E8%A8%AD%E5%AE%9A%E3%81%97%E3%81%A6%E3%81%84%E3%82%8Bkeyvalues

File: controllers/application_controller.go (hunk 2 above)
Comment (zoetrope) [reply controllers/application_controller.go:121]: `RequeAfter`: the backoff is constant.
`Reque`: the backoff is exponential.

I think `Reque` is more appropriate.

**Review by yokaze — APPROVED** [review]
LGTM

**Pull request merged**